src/main/java/net/minecraftforge/common/lighting/LightingEngine.java

```diff
@@ -0,0 +1,630 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.lighting;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.Minecraft;
+import net.minecraft.profiler.Profiler;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.BlockPos.MutableBlockPos;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.world.EnumSkyBlock;
+import net.minecraft.world.World;
+import net.minecraft.world.chunk.Chunk;
+
+public class LightingEngine
+{
+    private static final int MAX_SCHEDULED_COUNT = 1 << 22;
+
+    private static final int MAX_LIGHT = 15;
+
+    private static final Logger logger = LogManager.getLogger();
+
+    private final World world;
+    private final Profiler profiler;
+
+    //Layout of longs: [padding(4)] [z(26)] [y(8)] [x(26)]
+    private final PooledLongQueue[] queuedLightUpdates = new PooledLongQueue[EnumSkyBlock.values().length];
+
+    //Layout of longs: see above
+    private final PooledLongQueue[] queuedDarkenings = new PooledLongQueue[MAX_LIGHT + 1];
+    private final PooledLongQueue[] queuedBrightenings = new PooledLongQueue[MAX_LIGHT + 1];
+
+    //Layout of longs: [newLight(4)] [pos(60)]
+    private final PooledLongQueue initialBrightenings = new PooledLongQueue();
+    //Layout of longs: [padding(4)] [pos(60)]
+    private final PooledLongQueue initialDarkenings = new PooledLongQueue();
+
+    private boolean updating = false;
+
+    //Layout parameters
+    //Length of bit segments
+    private static final int
+        lX = 26,
+        lY = 8,
+        lZ = 26,
+        lL = 4;
+
+    //Big segment shifts/positions
+    private static final int
+        sX = 0,
+        sY = sX + lX,
+        sZ = sY + lY,
+        sL = sZ + lZ;
+
+    //Bit segment masks
+    private static final long
+        mX = (1 << lX) - 1,
+        mY = (1 << lY) - 1,
+        mZ = (1 << lZ) - 1,
+        mL = (1 << lL) - 1;
+
+    //Stored light type to reduce amount of method parameters
+    private EnumSkyBlock lightType;
+
+    //Iteration state data
+    //Cache position to avoid allocation of new object each time
+    private final MutableBlockPos curPos = new MutableBlockPos();
+    private PooledLongQueue curQueue;
+    private Chunk curChunk;
+    private long curData;
+
+    //Cached data about neighboring blocks (of tempPos)
+    private boolean isNeighborDataValid = false;
+    private final Chunk[] neighborsChunk = new Chunk[EnumFacing.VALUES.length];
+    private final MutableBlockPos[] neighborsPos = new MutableBlockPos[EnumFacing.VALUES.length];
+    private final int[] neighborsLight = new int[EnumFacing.VALUES.length];
+
```

Comment:
We could remove `lightType`, if that's preferred... 
Normally, we also prefer parameters, but for these, we decided to make an exception for the following reasons:
* We need the neighbor caching data, unless we pass some caching wrapper object back and forth, which is even uglier. The other alternative would be to inline `fetchNeighborData`
* The iteration state variables are needed for a similar reason (since we'd need to inline `nextItem` at 5 places)